base/abstractarray.jl

```diff
@@ -271,6 +271,114 @@ size_to_strides(s, d) = (s,)
 size_to_strides(s) = ()
 
 
+abstract type MemoryLayout{T} end
+struct UnknownLayout{T} <: MemoryLayout{T} end
+abstract type AbstractStridedLayout{T} <: MemoryLayout{T} end
+abstract type AbstractColumnMajor{T} <: AbstractStridedLayout{T} end
+struct DenseColumnMajor{T} <: AbstractColumnMajor{T} end
+struct ColumnMajor{T} <: AbstractColumnMajor{T} end
+abstract type AbstractRowMajor{T} <: AbstractStridedLayout{T} end
+struct DenseRowMajor{T} <: AbstractRowMajor{T} end
+struct RowMajor{T} <: AbstractRowMajor{T} end
+struct StridedLayout{T} <: AbstractStridedLayout{T} end
+
+"""
+    UnknownLayout{T}()
+
+is returned by `MemoryLayout(A)` if it is unknown how the entries of an array `A`
+are stored in memory.
+"""
+UnknownLayout
+
+"""
+    AbstractStridedLayout{T}
+
+is an abstract type whose subtypes are returned by `MemoryLayout(A)`
+if a matrix or vector `A` have storage laid out at regular offsets in memory,
+and which can therefore be passed to external C and Fortran functions expecting
+this memory layout.
+"""
+AbstractStridedLayout
+
+"""
+    DenseColumnMajor{T}()
+
+is returned by `MemoryLayout(A)` if a vector or matrix `A` has storage in memory
+equivalent to an `Array`, so that `stride(A,1) == 1` and `stride(A,2) == size(A,1)`.
+Arrays with `DenseColumnMajor` must conform to the `DenseArray` interface.
+"""
+DenseColumnMajor
+
+"""
+    ColumnMajor{T}()
+
+is returned by `MemoryLayout(A)` if a vector or matrix `A` has storage in memory
+as a column major matrix. In other words, the columns are stored in memory with
+offsets of one, while the rows are stored with offsets given by `stride(A,2)`.
+Arrays with `ColumnMajor` must conform to the `DenseArray` interface.
+"""
+ColumnMajor
+
+"""
+    DenseRowMajor{T}()
+
+is returned by `MemoryLayout(A)` if a matrix `A` has storage in memory
+equivalent to the transpose of an `Array`, so that `stride(A,1) == size(A,1)` and
+`stride(A,2) == 1`. Arrays with `DenseRowMajor` must conform to the
+`DenseArray` interface.
+"""
+DenseRowMajor
+
+"""
+    RowMajor{T}()
+
+is returned by `MemoryLayout(A)` if a matrix `A` has storage in memory
+as a row major matrix. In other words, the rows are stored in memory with
```

Comment:
Note that this documentation is specific to 2d arrays ("matrices"), but "row-major" and "column-major" are terms that have standard generalizations to higher dimensions.

Reply:
In particular, I would say that "row major" means that `stride(A,i) ≥ size(A,i+1) * stride(A,i+1)` (equality for dense row major) and `stride(A,ndims(A)) == 1`.

"Column major" is `stride(A,i+1) ≥ size(A,i) * stride(A,i)` with `stride(A,1) == 1`.

Reply:
👍 

I’ll update the docs.

I’ll probably mention matrices as a special case, since there it’s important for LAPACK